sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -250,15 +253,28 @@ object OptimizeSkewedJoin extends AQEShuffleReadRule {
       // SHJ
       //   Shuffle
       //   Shuffle
-      optimizeSkewJoin(plan)
+      val optimized =
+        EnsureRequirements(requiredDistribution.isDefined, requiredDistribution)
+          .apply(optimizeSkewJoin(plan))
+      val originCost = costEvaluator.evaluateCost(plan)
+      val optimizedCost = costEvaluator.evaluateCost(optimized)
+      // two cases we will pick new plan:
+      //   1. optimize the skew join without extra shuffle
+      //   2. optimize the skew join with extra shuffle but the costEvaluator think it's better
+      if (optimizedCost < originCost || (originCost == optimizedCost && optimized != plan)) {
```

Comment:
Moved the cost evaluation into `OptimizeSkewJoin`, so now we have two cases that will pick the new plan (see the comment).

Reply:
how about just `optimizedCost <= originCost`?

Reply:
yea, since we are only optimize physical plan here